### ISSUE TITLE
How to find supplementary materials

### ISSUE THREAD

**Issue by daidaiershidi** [MediaBrain-SJTU/EqMotion#1]
Thank you for bringing inspiring work. where can I find supplementary materials?

**Comment by sjtuxcx** [comment]
Thanks for your interest in our work! We have uploaded the supplementary materials to Arxiv: [link](https://arxiv.org/abs/2303.10876).

**Comment by daidaiershidi** [comment]
I am a beginner in this field. Could you recommend any introductory books?

**Comment by daidaiershidi** [comment]
I am a beginner in this field. Could you recommend any introductory books or papers?

**Comment by sjtuxcx** [comment]
Hi, you can refer to the related work section of our paper.

**Comment by daidaiershidi** [comment]
Hi, Equivariant inner-agent attention is implemented as follows:
```
    def inner_agent_attention(self,coord,h):
        att = self.mlp_q(h).unsqueeze(-1)
        v = coord - torch.mean(coord,dim=(1,2),keepdim=True)
        out = att * v
        apply_res = True
        if apply_res:
            out = out + coord
        return out
```

However, the equation in the paper seems to add the mean value:
![image](https://user-images.githubusercontent.com/48178838/229439786-efc36114-6e7f-4226-8d9e-6d88d808b365.png)

Does this break equivalence and what are the implications of doing so? @sjtuxcx

**Comment by daidaiershidi** [comment]
And non-linear is implemented as follows:
```
    def non_linear(self, coord):
        coord_mean = torch.mean(coord,dim=(1,2),keepdim=True)
        coord = coord - coord_mean
        q = self.layer_q(coord.transpose(2,3)).transpose(2,3)
        k = self.layer_k(coord.transpose(2,3)).transpose(2,3)
        product = torch.matmul(q.unsqueeze(-2),k.unsqueeze(-1)).squeeze(-1) # (B,N,C,1)
        mask = (product >= 0).float() # (B,N,C,1)
        EPS = 1e-4
        k_norm_sq = torch.sum(k*k,dim=-1,keepdim=True) # (B,N,C,1)
        coord = mask * q + (1-mask) * (q-(product/(k_norm_sq+EPS))*k)
        coord = coord + coord_mean
        return coord
```
However, the equation in the paper is:
![image](https://user-images.githubusercontent.com/48178838/229453199-169d7912-d35d-4e60-a80a-ee5afec1fbd4.png)

I think the corresponding implementation should be:

```
coord = mask * q + (1-mask) * (q-(product/(k_norm_sq+EPS))*(k/(k_norm_sq+EPS)))
```

**Comment by sjtuxcx** [comment]
Hi, sorry for the late reply, for your questions:
1. It will not break the equivariance. Here we apply a residual connection. You also can only add the coordinate mean (torch.mean(coord,dim=(1,2),keepdim=True)) instead.
2. The k_norm_sq is already squared of the k's L2 norm, thus you don't need to divide it twice.
